reinitialize-pods/src/lib.rs

```diff
@@ -0,0 +1,176 @@
+use futures_util::{Stream, StreamExt};
+use k8s_openapi::api::core::v1::Pod;
+use kube::{
+    runtime::{
+        events::{Event, EventType, Recorder, Reporter},
+        watcher,
+    },
+    Client, Error, Resource, ResourceExt,
+};
+use kubert::Runtime;
+use prometheus_client::{metrics::counter::Counter, registry::Registry};
+use tokio::sync::mpsc::{self, error::TrySendError, Receiver, Sender};
+
+// ERRNO 95: Operation not supported
+pub const UNSUCCESSFUL_EXIT_CODE: i32 = 95;
+
+// If the event channel capacity is reached, the event is dropped, but a new one will be emitted
+// in the pod's next crashloop iteration
+const EVENT_CHANNEL_CAPACITY: usize = 32;
+
+const DATA_PLANE_LABEL: &str = "linkerd.io/control-plane-ns";
+const CONDITION_EVICTED_REASON: &str = "EvictionByEvictionAPI";
+const EVENT_ACTION: &str = "Evicting";
+const EVENT_REASON: &str = "LinkerdCNINotConfigured";
+
+#[derive(Clone, Debug)]
+pub struct Metrics {
+    queue_overflow: Counter<u64>,
+    evicted_pods: Counter<u64>,
+}
+
+pub fn run(rt: &mut Runtime, node_name: String, controller_pod_name: String, metrics: Metrics) {
+    let pod_evts = rt.watch_all::<Pod>(
+        watcher::Config::default()
+            .labels(DATA_PLANE_LABEL)
+            .fields(&format!("spec.nodeName={node_name}")),
+    );
+    let (tx, rx) = mpsc::channel(EVENT_CHANNEL_CAPACITY);
+    tokio::spawn(process_events(pod_evts, tx, metrics.clone()));
+
+    let client = rt.client();
+    tokio::spawn(process_pods(client, controller_pod_name, rx, metrics));
```

Comment:
nit: have we considered just spawning a task inside of `process_events()` for each pod we want to process? Then they can all be run in parallel by the tokio runtime. Might eliminate the need for a channel and simplify the code a little bit?

e.g.

```rust
 if terminated && !evicted {
                let namespace = pod.namespace().unwrap();
                let name = pod.name_any();
                tokio::spawn(process_pod(client.clone(), pod_name, metrics.clone());
}
```

Both the metrics and client structs should be pretty cheap to clone.

Reply:
Thanks Matei, this triggered an interesting discussion (off-github). To summarize it here, spawning new tasks for each event would require us to also bound them somehow. The tokio runtime should provide us a way for doing this, but that can complicate things, so for now we're gonna keep with the channel-backed sequential approach and see how that performs.